tests/streaming_test.py

```diff
@@ -548,3 +548,61 @@ def test_StreamProcessor_clear() -> None:
 
     # Static values should be preserved after clear, so call_count remains 1
     assert make_static_a.call_count == 1
+
+
+def test_StreamProcessor_with_context() -> None:
+    Streamed = NewType('Streamed', int)
+    Context = NewType('Context', int)
+    Static = NewType('Static', int)
+    ProcessedContext = NewType('ProcessedContext', int)
+    ProcessedStreamed = NewType('ProcessedStreamed', int)
+    Output = NewType('Output', int)
+
+    def make_static() -> Static:
+        make_static.call_count += 1
+        return Static(2)
+
+    make_static.call_count = 0
+
+    def process_context(context: Context, static: Static) -> ProcessedContext:
+        process_context.call_count += 1
+        return ProcessedContext(context * static)
+
+    process_context.call_count = 0
+
+    def process_streamed(
+        streamed: Streamed, context: ProcessedContext
+    ) -> ProcessedStreamed:
+        return ProcessedStreamed(streamed + context)
+
+    def finalize(streamed: ProcessedStreamed) -> Output:
+        return Output(streamed)
+
+    wf = sciline.Pipeline((make_static, process_context, process_streamed, finalize))
+    streaming_wf = streaming.StreamProcessor(
+        base_workflow=wf,
+        dynamic_keys=(Streamed,),
+        context_keys=(Context,),
+        target_keys=(Output,),
+        accumulators=(ProcessedStreamed,),
+    )
+    assert make_static.call_count == 1
+    assert process_context.call_count == 0
+    streaming_wf.set_context({Context: sc.scalar(3)})
+    assert process_context.call_count == 1
```

Comment:
Not sure it is good that setting the context computes intermediates. This means that setting the context is expensive. Can this be done when streamed data is added?

I have two concerns about the current solution:
- Is there a possibility that `set_context` gets called while `accumulate` is running? If so, modifying the workflow while it is computing (more specifically, while it is building the task graph) may lead to inconsistent results. So this approach requires synchronisation between context changes and adding data to accumualte.
- The user of `StreamProcessor` might expect that only accumulating is expensive. So they may write code that a call to `set_context` is cheap. And this can create unexpected lags and difficult-to-understand performance problems.

Reply:
- StreamProcessor is not intended to be usable in a multi-threaded context, so not to the first question.
- I think this is only a concern if the context changes more often that the streamed data, which we probably need to to prevent in practice (filter noise?), or come up with a different solution. Regarding lag, it is actually important to to processing early, as soon as information is available. Say updating context takes 0.5 seconds, and we typically push streamed data twice per second. If we delay the context update until the next data push, you will experience the full 0.5 seconds lag. If you do it immediately some (unpredictable) amount of it can be hidden in the gap between streamed updates.

Reply:
This assumes that `T(context_update + stream_update) = T(context_update) + T(stream_update)`, where `T` is run time.
But it is fine, I just wanted to make sure you considered this.